File: packages/ts/src/discover.ts
/**
 * Search-path resolution — SPEC §2.
 *
 * Why this does not use `cosmiconfig.search()`: cosmiconfig's search walks upward and stops at
 * the **first** hit, returning one config. SPEC §2.2 collects **every** ancestor root and
 * layers them nearest-wins, which `search()` cannot express — and `strategy: "first-match"`
 * recovers the stop-at-first reading afterward from the full list. cosmiconfig is still doing
 * the job it is best at, one directory down: parsing and loading an individual file through its
 * loader registry (see loaders.ts). Anyone who sees the dependency and assumes `search()` is
 * driving discovery is reading it backwards.
 */

import { existsSync, realpathSync, statSync } from 'node:fs'
import { homedir } from 'node:os'
import { dirname, isAbsolute, join, resolve } from 'node:path'

import { ConfigError } from './errors.js'
import type { Format, WarningSink } from './types.js'

/** SPEC §2.5: the closed, ordered list of recognized file names. */
export const RECOGNIZED_FILES: ReadonlyArray<readonly [name: string, format: Format]> = [
  ['config.toml', 'toml'],
  ['config.yaml', 'yaml'],
  ['config.yml', 'yaml'],
  ['config.json', 'json'],
  ['config.jsonc', 'jsonc'],
  ['config.ini', 'ini'],
  ['.env', 'dotenv'],
]

/**
 * A pathological mount or a symlink loop the realpath did not collapse should fail loudly
 * rather than spin. No real tree is 64 directories deep below its repository root.
 */
const MAX_DEPTH = 64

export interface WalkOptions {
  /** The resolved home directory; the walk stops there, inclusive (SPEC §2.3). */
  home?: string
  /** Opt-in extra stop condition, inclusive. */
  stopDir?: string
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory()
  } catch {
    return false
  }
}

/**
 * Every existing `.config/<packageName>/` from `cwd` upward, **farthest ancestor first** so the
 * list is already in SPEC §2.7 order and the nearest root is last, and therefore wins.
 */
export function resolveProjectRoots(
  cwd: string,
  packageName: string,
  options: WalkOptions = {},
): string[] {
  let dir = realpathSync(resolve(cwd)) // SPEC §2.1: realpath-resolved exactly once.
  const home = options.home === undefined ? undefined : resolve(options.home)
  const stopDir = options.stopDir === undefined ? undefined : resolve(options.stopDir)

  const roots: string[] = []
  for (let depth = 0; ; depth++) {
    if (depth > MAX_DEPTH) {
      throw new ConfigError(
        'unreadable',
        `upward walk exceeded ${MAX_DEPTH} directories starting at ${cwd}; refusing to continue`,
        { path: cwd },
      )
    }

    // SPEC §2.2: a directory is checked before it is tested for stopping, so a config beside a
    // .git is found and the walk then ends.
    const candidate = join(dir, '.config', packageName)
    if (isDirectory(candidate)) roots.push(candidate)

    const parent = dirname(dir)
    const atFilesystemRoot = parent === dir
    const atHome = home !== undefined && dir === home
    const atStopDir = stopDir !== undefined && dir === stopDir
    // Both forms count: a directory in a normal clone, a file in a worktree or submodule.
    const atRepositoryBoundary = existsSync(join(dir, '.git'))
    if (atFilesystemRoot || atHome || atStopDir || atRepositoryBoundary) break

    dir = parent
  }
  return roots.reverse()
}

export interface UserRootOptions {
  /** Overrides the platform home directory. */
  home?: string
  /** Required — never read `process.env` in here; the probe and the tests both depend on that. */
  env: Record<string, string | undefined>
  onWarning?: WarningSink
}

/**
 * The single user-level root of SPEC §2.4, or `null` when it does not exist — so a caller can
 * tell "no user config at all" from "user config directory is empty".
 *
 * Windows takes this identical path. `%APPDATA%` and `%LOCALAPPDATA%` are deliberately not
 * consulted (SPEC §2.4): one documented location on every platform beats a native one that
 * nobody can predict from the docs, because the same directory has to be readable by five
 * language implementations.
 */
export function resolveUserRoot(packageName: string, options: UserRootOptions): string | null {
  const xdg = options.env['XDG_CONFIG_HOME']
  let root: string
  if (xdg !== undefined && xdg !== '' && isAbsolute(xdg)) {
    root = join(xdg, packageName)
  } else {
    if (xdg !== undefined) {
      options.onWarning?.(
        `ignoring XDG_CONFIG_HOME=${JSON.stringify(xdg)}: it must be a non-empty absolute path (SPEC §2.4)`,
      )
    }
    root = join(options.home ?? homedir(), '.config', packageName)
  }
  return isDirectory(root) ? root : null
}

export interface ConfigFileRef {
  path: string
  format: Format
}

/**
 * The recognized files present in one config directory, in SPEC §2.5 order — which is also
 * their load order, later entries winning.
 */
export function listConfigFiles(root: string, profile?: string): ConfigFileRef[] {
  // SPEC §2.5: a mistake, not an intention. Picking a winner silently would hide it.
  if (existsSync(join(root, 'config.yaml')) && existsSync(join(root, 'config.yml'))) {
    throw new ConfigError(
      'duplicate-format',
      `${root}: config.yaml and config.yml cannot both be present`,
      { path: root },
    )
  }

  const files: ConfigFileRef[] = []
  for (const [name, format] of RECOGNIZED_FILES) {
    const path = join(root, name)
    if (existsSync(path)) files.push({ path, format })
    if (profile !== undefined && profile !== '') {
      // SPEC §2.6: `config.<profile>.<ext>` immediately after its base file.
      const dot = name.lastIndexOf('.')
      const profiled = dot <= 0 ? `${name}.${profile}` : `${name.slice(0, dot)}.${profile}${name.slice(dot)}`
      const profiledPath = join(root, profiled)
      if (existsSync(profiledPath)) files.push({ path: profiledPath, format })
    }
  }
  return files
}

File: packages/golang/discover.go
package configdiscovery

// Search-path resolution - SPEC section 2.
//
// Viper's AddConfigPath searches a flat list of directories and stops at the first match; it
// never walks upward, and it cannot express "collect every ancestor and layer them". So the walk
// is entirely ours, in every language, and this file is the Go half of keeping those five
// hand-written walks identical.

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// maxDepth guards against a pathological mount or an uncollapsed symlink loop. No real tree is
// 64 directories deep below its repository root.
const maxDepth = 64

type walkOptions struct {
	home    string
	stopDir string
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// resolveProjectRoots returns every existing .config/<packageName>/ from cwd upward, farthest
// ancestor first - the SPEC section 2.7 order, so the nearest root is last and therefore wins.
func resolveProjectRoots(cwd, packageName string, o walkOptions) ([]string, error) {
	// SPEC section 2.1: resolved exactly once. A failure here is reported rather than papered
	// over, because silently walking an unresolved path would search the wrong ancestors.
	dir, err := filepath.EvalSymlinks(cwd)
	if err != nil {
		return nil, &ConfigError{Kind: KindUnreadable, Path: cwd, Message: err.Error(), Err: err}
	}

	home := resolveOrRaw(o.home)
	stopDir := resolveOrRaw(o.stopDir)

	var roots []string
	for depth := 0; ; depth++ {
		if depth > maxDepth {
			return nil, &ConfigError{
				Kind:    KindUnreadable,
				Path:    cwd,
				Message: fmt.Sprintf("upward walk exceeded %d directories", maxDepth),
			}
		}

		// SPEC section 2.2: a directory is checked before it is tested for stopping, so a config
		// beside a .git is found and the walk then ends.
		if candidate := filepath.Join(dir, ".config", packageName); isDir(candidate) {
			roots = append(roots, candidate)
		}

		parent := filepath.Dir(dir)
		atFilesystemRoot := parent == dir
		atHome := home != "" && dir == home
		atStopDir := stopDir != "" && dir == stopDir
		// Lstat, and both forms count: a directory in a normal clone, a file in a worktree or a
		// submodule. Those are repositories too.
		_, gitErr := os.Lstat(filepath.Join(dir, ".git"))
		atRepositoryBoundary := gitErr == nil

		if atFilesystemRoot || atHome || atStopDir || atRepositoryBoundary {
			break
		}
		dir = parent
	}

	for i, j := 0, len(roots)-1; i < j; i, j = i+1, j-1 {
		roots[i], roots[j] = roots[j], roots[i]
	}
	return roots, nil
}

// resolveOrRaw resolves symlinks when it can and falls back to the input otherwise, so a stop
// directory that does not exist yet still compares against the walk by name.
func resolveOrRaw(path string) string {
	if path == "" {
		return ""
	}
	if resolved, err := filepath.EvalSymlinks(path); err == nil {
		return resolved
	}
	return filepath.Clean(path)
}

// resolveUserRoot returns the single user-level root of SPEC section 2.4, or "" when there is
// none - so a caller can tell "no user config at all" from "the user config directory is empty".
//
// Windows takes this identical path. %APPDATA% and %LOCALAPPDATA% are deliberately not consulted
// (SPEC section 2.4): the same directory has to be readable by five language implementations,
// and one documented location beats a native one nobody can predict.
func resolveUserRoot(packageName, home string, env map[string]string, warn func(string)) string {
	var root string
	xdg, set := env["XDG_CONFIG_HOME"]
	if set && xdg != "" && filepath.IsAbs(xdg) {
		root = filepath.Join(xdg, packageName)
	} else {
		if set {
			warn(fmt.Sprintf(
				"ignoring XDG_CONFIG_HOME=%q: it must be a non-empty absolute path (SPEC section 2.4)",
				xdg,
			))
		}
		root = filepath.Join(home, ".config", packageName)
	}
	if !isDir(root) {
		return ""
	}
	return root
}

type fileRef struct {
	path   string
	format string
}

// recognizedFiles is the closed, ordered list of SPEC section 2.5. That order is also the load
// order within one directory, later entries winning.
var recognizedFiles = []fileRef{
	{"config.toml", "toml"},
	{"config.yaml", "yaml"},
	{"config.yml", "yaml"},
	{"config.json", "json"},
	{"config.jsonc", "jsonc"},
	{"config.ini", "ini"},
	{".env", "dotenv"},
}

func exists(path string) bool {
	_, err := os.Lstat(path)
	return err == nil
}

// listConfigFiles returns the recognized files present in one config directory, in SPEC section
// 2.5 order.
func listConfigFiles(root, profile string) ([]fileRef, error) {
	// SPEC section 2.5: a mistake, not an intention. Picking a winner silently would hide it.
	if exists(filepath.Join(root, "config.yaml")) && exists(filepath.Join(root, "config.yml")) {
		return nil, &ConfigError{
			Kind:    KindDuplicateFormat,
			Path:    root,
			Message: "config.yaml and config.yml cannot both be present",
		}
	}

	var files []fileRef
	for _, candidate := range recognizedFiles {
		path := filepath.Join(root, candidate.name())
		if exists(path) {
			files = append(files, fileRef{path: path, format: candidate.format})
		}
		if profile != "" {
			// SPEC section 2.6: config.<profile>.<ext> immediately after its base file.
			profiled := filepath.Join(root, withProfile(candidate.name(), profile))
			if exists(profiled) {
				files = append(files, fileRef{path: profiled, format: candidate.format})
			}
		}
	}
	return files, nil
}

// name is the base file name for an entry of recognizedFiles, where path holds the name rather
// than a full path.
func (f fileRef) name() string { return f.path }

func withProfile(name, profile string) string {
	dot := strings.LastIndex(name, ".")
	if dot <= 0 {
		return name + "." + profile
	}
	return name[:dot] + "." + profile + name[dot:]
}
